File: server/app/routes/agent_runtimes.py
from fastapi import APIRouter

from server.app.routes.agent_runtimes_contracts import AgentRuntimesResponse
from server.app.services.agent_runtime_catalog import AgentRuntimeCatalogService


def create_agent_runtimes_router(
    catalog: AgentRuntimeCatalogService | None = None,
) -> APIRouter:
    """Per-runtime tool catalog for Studio's dynamic tool picker (#476).

    Read-only projection of the runtime adapter declarations — no DB access,
    so the service is stateless and lazily constructed.
    """
    router = APIRouter()

    @router.get(
        "/agent-runtimes",
        response_model=AgentRuntimesResponse,
        response_model_exclude_none=True,
    )
    def get_agent_runtimes() -> AgentRuntimesResponse:
        service = catalog or AgentRuntimeCatalogService()
        return AgentRuntimesResponse(**service.tool_catalog())

    return router

File: frontend/src/api/agentRuntimes.ts
import { api } from './core'
import type { AgentRuntimesResponse } from '../types'

/**
 * #476：per-runtime 工具目录（按 runtime 嵌套）。同名工具交集不是契约
 * ——description/parameters 随 runtime 走，消费方不得借交集建立跨
 * runtime 统一语义。目录数据与 dispatch 期工具名校验（#449）同源。
 */
export const fetchAgentRuntimes = () =>
  api<AgentRuntimesResponse>('/api/agent-runtimes')
